### PR TITLE
Custom Fields/Types, closes #593, closes #644

### LINKED ISSUE #593: add experimental feature custom types
* http://dev.sphere.io/dev/http-api-projects-custom-fields.html
* http://dev.sphere.io/dev/http-api-projects-types.html

* [x] Type
* [x] Field Definition
* [x] FieldType
    * [x] BooleanType
    * [x] StringType
    * [x] LocalizedStringType
    * [x] EnumType
    * [x] LocalizedEnumType
    * [x] NumberType
    * [x] MoneyType
    * [x] DateType
    * [x] TimeType
    * [x] DateTimeType
    * [x] ReferenceType
    * [x] SetType
* [x] TextInputHint to common
* [x] EnumValue
* [x] LocalizedEnumValue
* [x] Get Type by ID
* [x] Query Types
* [x] Create a Type
* [x] Update a Type
    * [x] changeName
    * [x] setDescription
    * [x] addFieldDefinition
    * [x] removeFieldDefinition
    * [x] changeLabel
    * [x] addEnumValue
    * [x] addLocalizedEnumValue
    * [ ] changeFieldDefinitionOrder
    * [ ] changeEnumValueOrder
    * [ ] changeLocalizedEnumValueOrder
* [x] delete
* [x] Custom
* [x] CustomField
* [ ] New Custom Fields = CustomFieldsDraft?
* [x] custom field in Categories (add to draft object/builder)
* [x] custom field update actions in Categories
* [x] custom field in Customers (add to draft object/builder)
* [x] custom field update actions in Customers
* [x] custom field in Carts (add to draft object/builder)
* [x] custom field update actions in Carts
* [x] custom field in Orders (add to draft object/builder)
* [x] custom field update actions in Orders
* [x] custom field in line items (add to draft object/builder)
* [x] custom field update actions in line items, cart and order
* [x] custom field in custom line items (add to draft object/builder)
* [x] custom field update actions in line items, cart and order
* [x] required validation (required field(
* [x] add interface Custom to the types in http://dev.sphere.io/http-api-projects-custom-fields.html#customizable-resource
* [x] check reference expansion, may update query and expansion models, see https://github.com/sphereio/sphere-jvm-sdk/issues/730
* [x] test io.sphere.sdk.types.CustomFieldsDraftBuilder#ofTypeKey

### LINKED ISSUE #644: Do not log char array from responses
```
11:18:12.755 [ForkJoinPool-1-worker-1] DEBUG s.p.responses.queries - HttpResponseImpl[statusCode=200,headers={Transfer-Encoding=[chunked], Server=[nginx], Access-Control-Allow-Origin=[*], X-Correlation-ID=[nginx-c2824e4d-ee11-426c-ad89-b3b370b71a41], X-Served-By=[app08.sphere.prod.commercetools.de], Access-Control-Allow-Methods=[GET, POST, DELETE, OPTIONS], Connection=[keep-alive], X-Served-Config=[sphere-projects-ws-1.0], Date=[Tue, 11 Aug 2015 09:17:50 GMT], Access-Control-Allow-Headers=[Accept, Authorization, Content-Type, Origin], Content-Type=[application/json; charset=utf-8]},responseBody={123,34,111,102,102,115,101,116,34,58,48,44,34,99,111,117,110,116,34,58,53,44,34,116,111,116,97,108,34,58,53,44,34,114,101,115,117,108,116,115,34,58,91,123,34,105,100,34,58,34,51,51,52,102,48,57,55,100,45,56,57,49,56,45,52,50,48,102,45,98,54,102,52,45,97,98,56,54,54,57,54,99,49,102,54,55,34,44,34,118,101,114,115,105,111,110,34,58,50,44,34,112,114,111,100,117,99,116,84,121,112,101,34,58,123,34,116,121,112,101,73,100,34,58,34,112,114,111,100,117,99,116,45,116,121,112,101,34,44,34,105,100,34,58,34,101,53,57,98,50,98,101,54,45,56,99,48,100,45,52,56,53,102,45,56,102,100,51,45,55,56,53,49,48,57,100,57,56,48,57,49,34,125,44,34,110,97,109,101,34,58,123,34,101,110,34,58,34,83,65,80,80,72,73,82,69,34,125,44,34,100,101,115,99,114,105,112,116,105,111,110,34,58,123,34,101,110,34,58,34,83,97,109,112,108,101,32,100,101,115,99,114,105,112,116,105,111,110,34,125,44,34,99,97,116,101,103,111,114,105,101,115,34,58,91,123,34,116,121,112,101,73,100,34,58,34,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,48,53,102,49,97,50,57,102,45,97,100,54,99,45,52,55,51,51,45,56,98,57,53,45,98,100,56,51,97,49,98,100,100,57,54,54,34,125,93,44,34,115,108,117,103,34,58,123,34,101,110,34,58,34,115,97,112,112,104,105,114,101,49,52,51,56,57,52,56,51,52,53,57,49,56,34,125,44,34,109,97,115,116,101,114,86,97,114,105,97,110,116,34,58,123,34,105,100,34,58,49,44,34,115,107,117,34,58,34,115,107,117,95,83,65,80,80,72,73,82,69,95,118,97,114,105,97,110,116,49,95,49,52,51,56,57,52,56,51,52,53,57,49,56,34,44,34,112,114,105,99,101,115,34,58,91,123,34,118,97,108,117,101,34,58,123,34,99,117,114,114,101,110,99,121,67,111,100,101,34,58,34,69,85,82,34,44,34,99,101,110,116,65,109,111,117,110,116,34,58,50,56,48,48,125,44,34,105,100,34,58,34,56,56,52,101,52,49,53,52,45,100,51,49,52,45,52,52,98,99,45,97,57,100,98,45,102,57,51,55,50,50,49,102,55,102,57,49,34,125,93,44,34,105,109,97,103,101,115,34,58,91,123,34,117,114,108,34,58,34,104,116,116,112,115,58,47,47,119,119,119,46,99,111,109,109,101,114,99,101,116,111,111,108,115,46,99,111,109,47,99,108,105,47,100,97,116,97,47,50,53,50,53,52,50,48,48,53,95,49,46,106,112,103,34,44,34,100,105,109,101,110,115,105,111,110,115,34,58,123,34,119,34,58,49,52,48,48,44,34,104,34,58,49,52,48,48,125,125,93,44,34,97,116,116,114,105,98,117,116,101,115,34,58,91,93,125,44,34,118,97,114,105,97,110,116,115,34,58,91,93,44,34,115,101,97,114,99,104,75,101,121,119,111,114,100,115,34,58,123,125,44,34,104,97,115,83,116,97,103,101,100,67,104,97,110,103,101,115,34,58,102,97,108,115,101,44,34,112,117,98,108,105,115,104,101,100,34,58,116,114,117,101,44,34,116,97,120,67,97,116,101,103,111,114,121,34,58,123,34,116,121,112,101,73,100,34,58,34,116,97,120,45,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,52,57,51,55,55,55,52,98,45,102,102,53,98,45,52,54,98,98,45,97,48,102,55,45,54,102,54,52,51,55,55,48,99,53,55,50,34,125,44,34,99,114,101,97,116,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,53,46,57,52,56,90,34,44,34,108,97,115,116,77,111,100,105,102,105,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,50,55,90,34,125,44,123,34,105,100,34,58,34,97,100,55,99,98,49,97,55,45,57,98,54,53,45,52,97,51,51,45,98,51,100,51,45,97,56,55,101,98,97,56,52,57,55,51,100,34,44,34,118,101,114,115,105,111,110,34,58,50,44,34,112,114,111,100,117,99,116,84,121,112,101,34,58,123,34,116,121,112,101,73,100,34,58,34,112,114,111,100,117,99,116,45,116,121,112,101,34,44,34,105,100,34,58,34,101,53,57,98,50,98,101,54,45,56,99,48,100,45,52,56,53,102,45,56,102,100,51,45,55,56,53,49,48,57,100,57,56,48,57,49,34,125,44,34,110,97,109,101,34,58,123,34,101,110,34,58,34,87,66,32,65,84,72,76,69,84,73,67,32,84,65,78,75,34,125,44,34,100,101,115,99,114,105,112,116,105,111,110,34,58,123,34,101,110,34,58,34,83,97,109,112,108,101,32,100,101,115,99,114,105,112,116,105,111,110,34,125,44,34,99,97,116,101,103,111,114,105,101,115,34,58,91,123,34,116,121,112,101,73,100,34,58,34,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,56,48,99,98,51,98,99,50,45,57,52,101,52,45,52,99,51,52,45,57,54,54,100,45,56,52,100,97,54,51,98,97,99,50,57,102,34,125,93,44,34,115,108,117,103,34,58,123,34,101,110,34,58,34,119,98,45,97,116,104,108,101,116,105,99,45,116,97,110,107,49,52,51,56,57,52,56,51,52,53,57,54,50,34,125,44,34,109,97,115,116,101,114,86,97,114,105,97,110,116,34,58,123,34,105,100,34,58,49,44,34,115,107,117,34,58,34,115,107,117,95,87,66,95,65,84,72,76,69,84,73,67,95,84,65,78,75,95,118,97,114,105,97,110,116,49,95,49,52,51,56,57,52,56,51,52,53,57,54,50,34,44,34,112,114,105,99,101,115,34,58,91,123,34,118,97,108,117,101,34,58,123,34,99,117,114,114,101,110,99,121,67,111,100,101,34,58,34,69,85,82,34,44,34,99,101,110,116,65,109,111,117,110,116,34,58,56,52,48,48,125,44,34,105,100,34,58,34,49,54,98,100,49,48,102,102,45,98,99,56,102,45,52,48,55,56,45,98,55,56,52,45,101,100,50,54,52,50,100,52,97,52,99,53,34,125,93,44,34,105,109,97,103,101,115,34,58,91,123,34,117,114,108,34,58,34,104,116,116,112,115,58,47,47,119,119,119,46,99,111,109,109,101,114,99,101,116,111,111,108,115,46,99,111,109,47,99,108,105,47,100,97,116,97,47,50,53,51,50,54,53,52,52,52,95,49,46,106,112,103,34,44,34,100,105,109,101,110,115,105,111,110,115,34,58,123,34,119,34,58,49,52,48,48,44,34,104,34,58,49,52,48,48,125,125,93,44,34,97,116,116,114,105,98,117,116,101,115,34,58,91,93,125,44,34,118,97,114,105,97,110,116,115,34,58,91,93,44,34,115,101,97,114,99,104,75,101,121,119,111,114,100,115,34,58,123,125,44,34,104,97,115,83,116,97,103,101,100,67,104,97,110,103,101,115,34,58,102,97,108,115,101,44,34,112,117,98,108,105,115,104,101,100,34,58,116,114,117,101,44,34,116,97,120,67,97,116,101,103,111,114,121,34,58,123,34,116,121,112,101,73,100,34,58,34,116,97,120,45,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,52,57,51,55,55,55,52,98,45,102,102,53,98,45,52,54,98,98,45,97,48,102,55,45,54,102,54,52,51,55,55,48,99,53,55,50,34,125,44,34,99,114,101,97,116,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,53,46,57,54,57,90,34,44,34,108,97,115,116,77,111,100,105,102,105,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,52,54,90,34,125,44,123,34,105,100,34,58,34,56,54,52,55,102,53,55,98,45,53,51,102,49,45,52,55,55,54,45,98,48,53,97,45,53,100,48,52,57,51,57,48,48,56,98,101,34,44,34,118,101,114,115,105,111,110,34,58,50,44,34,112,114,111,100,117,99,116,84,121,112,101,34,58,123,34,116,121,112,101,73,100,34,58,34,112,114,111,100,117,99,116,45,116,121,112,101,34,44,34,105,100,34,58,34,101,53,57,98,50,98,101,54,45,56,99,48,100,45,52,56,53,102,45,56,102,100,51,45,55,56,53,49,48,57,100,57,56,48,57,49,34,125,44,34,110,97,109,101,34,58,123,34,101,110,34,58,34,71,73,82,76,83,32,72,65,82,84,66,82,69,65,75,32,67,82,69,87,34,125,44,34,100,101,115,99,114,105,112,116,105,111,110,34,58,123,34,101,110,34,58,34,83,97,109,112,108,101,32,100,101,115,99,114,105,112,116,105,111,110,34,125,44,34,99,97,116,101,103,111,114,105,101,115,34,58,91,123,34,116,121,112,101,73,100,34,58,34,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,55,48,52,56,100,99,97,49,45,98,52,99,99,45,52,51,55,97,45,98,54,55,100,45,48,98,55,51,50,99,101,98,55,49,98,99,34,125,93,44,34,115,108,117,103,34,58,123,34,101,110,34,58,34,103,105,114,108,115,45,104,97,114,116,98,114,101,97,107,45,99,114,101,119,49,52,51,56,57,52,56,51,52,53,57,53,49,34,125,44,34,109,97,115,116,101,114,86,97,114,105,97,110,116,34,58,123,34,105,100,34,58,49,44,34,115,107,117,34,58,34,115,107,117,95,71,73,82,76,83,95,72,65,82,84,66,82,69,65,75,95,67,82,69,87,95,118,97,114,105,97,110,116,49,95,49,52,51,56,57,52,56,51,52,53,57,53,49,34,44,34,112,114,105,99,101,115,34,58,91,123,34,118,97,108,117,101,34,58,123,34,99,117,114,114,101,110,99,121,67,111,100,101,34,58,34,69,85,82,34,44,34,99,101,110,116,65,109,111,117,110,116,34,58,51,52,48,48,125,44,34,105,100,34,58,34,100,101,49,50,98,55,54,55,45,48,49,56,50,45,52,101,49,48,45,56,101,53,101,45,51,57,48,101,98,97,98,57,53,101,100,101,34,125,93,44,34,105,109,97,103,101,115,34,58,91,123,34,117,114,108,34,58,34,104,116,116,112,115,58,47,47,119,119,119,46,99,111,109,109,101,114,99,101,116,111,111,108,115,46,99,111,109,47,99,108,105,47,100,97,116,97,47,50,53,51,50,51,52,51,56,55,95,49,46,106,112,103,34,44,34,100,105,109,101,110,115,105,111,110,115,34,58,123,34,119,34,58,49,52,48,48,44,34,104,34,58,49,52,48,48,125,125,93,44,34,97,116,116,114,105,98,117,116,101,115,34,58,91,93,125,44,34,118,97,114,105,97,110,116,115,34,58,91,93,44,34,115,101,97,114,99,104,75,101,121,119,111,114,100,115,34,58,123,125,44,34,104,97,115,83,116,97,103,101,100,67,104,97,110,103,101,115,34,58,102,97,108,115,101,44,34,112,117,98,108,105,115,104,101,100,34,58,116,114,117,101,44,34,116,97,120,67,97,116,101,103,111,114,121,34,58,123,34,116,121,112,101,73,100,34,58,34,116,97,120,45,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,52,57,51,55,55,55,52,98,45,102,102,53,98,45,52,54,98,98,45,97,48,102,55,45,54,102,54,52,51,55,55,48,99,53,55,50,34,125,44,34,99,114,101,97,116,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,53,46,57,53,57,90,34,44,34,108,97,115,116,77,111,100,105,102,105,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,51,54,90,34,125,44,123,34,105,100,34,58,34,99,55,50,98,51,56,98,48,45,51,54,101,53,45,52,98,54,97,45,97,102,51,97,45,99,50,57,53,98,52,54,51,57,55,101,98,34,44,34,118,101,114,115,105,111,110,34,58,50,44,34,112,114,111,100,117,99,116,84,121,112,101,34,58,123,34,116,121,112,101,73,100,34,58,34,112,114,111,100,117,99,116,45,116,121,112,101,34,44,34,105,100,34,58,34,101,53,57,98,50,98,101,54,45,56,99,48,100,45,52,56,53,102,45,56,102,100,51,45,55,56,53,49,48,57,100,57,56,48,57,49,34,125,44,34,110,97,109,101,34,58,123,34,101,110,34,58,34,77,66,32,83,79,70,84,83,72,69,76,76,32,76,73,78,69,82,34,125,44,34,100,101,115,99,114,105,112,116,105,111,110,34,58,123,34,101,110,34,58,34,83,97,109,112,108,101,32,100,101,115,99,114,105,112,116,105,111,110,34,125,44,34,99,97,116,101,103,111,114,105,101,115,34,58,91,123,34,116,121,112,101,73,100,34,58,34,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,48,53,102,49,97,50,57,102,45,97,100,54,99,45,52,55,51,51,45,56,98,57,53,45,98,100,56,51,97,49,98,100,100,57,54,54,34,125,93,44,34,115,108,117,103,34,58,123,34,101,110,34,58,34,109,98,45,115,111,102,116,115,104,101,108,108,45,108,105,110,101,114,49,52,51,56,57,52,56,51,52,53,57,57,52,34,125,44,34,109,97,115,116,101,114,86,97,114,105,97,110,116,34,58,123,34,105,100,34,58,49,44,34,115,107,117,34,58,34,115,107,117,95,77,66,95,83,79,70,84,83,72,69,76,76,95,76,73,78,69,82,95,118,97,114,105,97,110,116,49,95,49,52,51,56,57,52,56,51,52,53,57,57,52,34,44,34,112,114,105,99,101,115,34,58,91,123,34,118,97,108,117,101,34,58,123,34,99,117,114,114,101,110,99,121,67,111,100,101,34,58,34,69,85,82,34,44,34,99,101,110,116,65,109,111,117,110,116,34,58,49,48,48,48,48,125,44,34,105,100,34,58,34,99,49,51,100,99,52,50,53,45,101,102,48,102,45,52,53,97,55,45,56,50,54,99,45,102,52,49,98,98,99,52,51,102,102,52,55,34,125,93,44,34,105,109,97,103,101,115,34,58,91,123,34,117,114,108,34,58,34,104,116,116,112,115,58,47,47,119,119,119,46,99,111,109,109,101,114,99,101,116,111,111,108,115,46,99,111,109,47,99,108,105,47,100,97,116,97,47,50,53,52,51,57,49,54,51,49,95,49,46,106,112,103,34,44,34,100,105,109,101,110,115,105,111,110,115,34,58,123,34,119,34,58,49,52,48,48,44,34,104,34,58,49,52,48,48,125,125,93,44,34,97,116,116,114,105,98,117,116,101,115,34,58,91,93,125,44,34,118,97,114,105,97,110,116,115,34,58,91,93,44,34,115,101,97,114,99,104,75,101,121,119,111,114,100,115,34,58,123,125,44,34,104,97,115,83,116,97,103,101,100,67,104,97,110,103,101,115,34,58,102,97,108,115,101,44,34,112,117,98,108,105,115,104,101,100,34,58,116,114,117,101,44,34,116,97,120,67,97,116,101,103,111,114,121,34,58,123,34,116,121,112,101,73,100,34,58,34,116,97,120,45,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,52,57,51,55,55,55,52,98,45,102,102,53,98,45,52,54,98,98,45,97,48,102,55,45,54,102,54,52,51,55,55,48,99,53,55,50,34,125,44,34,99,114,101,97,116,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,48,49,90,34,44,34,108,97,115,116,77,111,100,105,102,105,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,53,55,90,34,125,44,123,34,105,100,34,58,34,99,48,52,97,102,49,52,100,45,54,57,48,55,45,52,49,100,52,45,98,57,98,49,45,48,54,101,56,49,54,49,51,53,50,49,101,34,44,34,118,101,114,115,105,111,110,34,58,50,44,34,112,114,111,100,117,99,116,84,121,112,101,34,58,123,34,116,121,112,101,73,100,34,58,34,112,114,111,100,117,99,116,45,116,121,112,101,34,44,34,105,100,34,58,34,101,53,57,98,50,98,101,54,45,56,99,48,100,45,52,56,53,102,45,56,102,100,51,45,55,56,53,49,48,57,100,57,56,48,57,49,34,125,44,34,110,97,109,101,34,58,123,34,101,110,34,58,34,77,66,32,80,82,69,77,73,85,77,32,84,69,67,72,32,84,34,125,44,34,100,101,115,99,114,105,112,116,105,111,110,34,58,123,34,101,110,34,58,34,83,97,109,112,108,101,32,100,101,115,99,114,105,112,116,105,111,110,34,125,44,34,99,97,116,101,103,111,114,105,101,115,34,58,91,123,34,116,121,112,101,73,100,34,58,34,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,55,48,52,56,100,99,97,49,45,98,52,99,99,45,52,51,55,97,45,98,54,55,100,45,48,98,55,51,50,99,101,98,55,49,98,99,34,125,93,44,34,115,108,117,103,34,58,123,34,101,110,34,58,34,109,98,45,112,114,101,109,105,117,109,45,116,101,99,104,45,116,49,52,51,56,57,52,56,51,52,54,48,48,57,34,125,44,34,109,97,115,116,101,114,86,97,114,105,97,110,116,34,58,123,34,105,100,34,58,49,44,34,115,107,117,34,58,34,115,107,117,95,77,66,95,80,82,69,77,73,85,77,95,84,69,67,72,95,84,95,118,97,114,105,97,110,116,49,95,49,52,51,56,57,52,56,51,52,54,48,48,57,34,44,34,112,114,105,99,101,115,34,58,91,123,34,118,97,108,117,101,34,58,123,34,99,117,114,114,101,110,99,121,67,111,100,101,34,58,34,69,85,82,34,44,34,99,101,110,116,65,109,111,117,110,116,34,58,49,48,48,48,48,125,44,34,105,100,34,58,34,53,57,53,49,48,56,97,52,45,55,100,48,53,45,52,50,51,48,45,97,100,99,57,45,49,52,97,102,102,49,51,56,53,99,102,56,34,125,93,44,34,105,109,97,103,101,115,34,58,91,123,34,117,114,108,34,58,34,104,116,116,112,115,58,47,47,119,119,119,46,99,111,109,109,101,114,99,101,116,111,111,108,115,46,99,111,109,47,99,108,105,47,100,97,116,97,47,50,53,51,50,52,53,56,50,49,95,49,46,106,112,103,34,44,34,100,105,109,101,110,115,105,111,110,115,34,58,123,34,119,34,58,49,52,48,48,44,34,104,34,58,49,52,48,48,125,125,93,44,34,97,116,116,114,105,98,117,116,101,115,34,58,91,93,125,44,34,118,97,114,105,97,110,116,115,34,58,91,93,44,34,115,101,97,114,99,104,75,101,121,119,111,114,100,115,34,58,123,125,44,34,104,97,115,83,116,97,103,101,100,67,104,97,110,103,101,115,34,58,102,97,108,115,101,44,34,112,117,98,108,105,115,104,101,100,34,58,116,114,117,101,44,34,116,97,120,67,97,116,101,103,111,114,121,34,58,123,34,116,121,112,101,73,100,34,58,34,116,97,120,45,99,97,116,101,103,111,114,121,34,44,34,105,100,34,58,34,52,57,51,55,55,55,52,98,45,102,102,53,98,45,52,54,98,98,45,97,48,102,55,45,54,102,54,52,51,55,55,48,99,53,55,50,34,125,44,34,99,114,101,97,116,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,49,55,90,34,44,34,108,97,115,116,77,111,100,105,102,105,101,100,65,116,34,58,34,50,48,49,53,45,48,56,45,48,55,84,49,49,58,53,50,58,50,54,46,48,54,52,90,34,125,93,125},associatedRequest=HttpRequestImpl[httpMethod=GET,url=https://api.sphere.io/hello-87/product-projections?staged=false,headers={User-Agent=[SPHERE.IO JVM SDK 1.0.0-M16], Authorization=[**removed from output**]},body=<null>],bodyAsStringForDebugging=<null>]
ProductProjectionImpl[productType=Reference{typeId='product-type', id='e59b2be6-8c0d-485f-8fd3-785109d98091', obj=null},taxCategory=Reference{typeId='tax-category', id='4937774b-ff5b-46bb-a0f7-6f643770c572', obj=null},isPublished=true,hasStagedChanges=false,name=LocalizedString(en -> SAPPHIRE),categories=[Reference{typeId='category', id='05f1a29f-ad6c-4733-8b95-bd83a1bdd966', obj=null}],description=LocalizedString(en -> Sample description),slug=LocalizedString(en -> sapphire1438948345918),metaTitle=<null>,metaDescription=<null>,metaKeywords=<null>,masterVariant=ProductVariantImpl[productId=334f097d-8918-420f-b6f4-ab86696c1f67,id=1,sku=sku_SAPPHIRE_variant1_1438948345918,prices=[Price[value=MoneyImpl[money=EUR 28.00000],country=<null>,customerGroup=<null>,channel=<null>,discounted=<null>,validFrom=<null>,validUntil=<null>,id=884e4154-d314-44bc-a9db-f937221f7f91]],images=[ImageImpl[url=https://www.commercetools.com/cli/data/252542005_1.jpg,dimensions=ImageDimensions[width=1400,height=1400],label=<null>]],availability=<null>,attributes=[]],variants=[],searchKeywords=SearchKeywords[content={}],id=334f097d-8918-420f-b6f4-ab86696c1f67,version=2,createdAt=2015-08-07T11:52:25.948Z,lastModifiedAt=2015-08-07T11:52:26.027Z]


```

### PR DESCRIPTION
**Pull request by schleichardt**

### CODE REVIEW

**Comment by schleichardt** [comment]
@lauraluiz @yanns please review.

The travis build will be red. They would be fixed by merging the master into the branch but then would be the review more difficult.

File: sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        this.fields = fields;
+    }
+
+    @Nullable
+    public JsonNode getField(final String name) {
+        return fields.get(name);
+    }
+
+    @Nullable
+    public <T> T getField(final String name, final TypeReference<T> typeReference) {
```
Comment (yanns) [inline sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java]: Should we add some helpers here, like:
```
public String getFieldAsString(final String name) {
    return getField(name, TypeReferences.stringTypeReference())
}

public Money getFieldAsMoney(final String name) 
...
```

**Comment by yanns** [comment]
great work! :+1: 

Later, if we need it, we could add a `TypedCustomFields` that take `CustomFields` and a `Type` as parameters and produces a Map where all custom fields are typed as String, Money... (either at construction time or lazily)
That could be helpful for a generic shop like sunrise.
Actually, we could try to build it in sunrise and later move it to the SDK if we feel confident about it.

File: sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java (hunk 1 above)
Comment (schleichardt) [inline sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java]: Yes, I will add them in this pull request.

Comment (schleichardt) [inline sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java]: I added the helpers. It is great in the custom types, that the enum types work symmetric concerning read and write. You set a string and get a string (key).

Comment (yanns) [inline sphere-models/src/main/java/io/sphere/sdk/types/CustomFields.java]: :+1:

File: sphere-models/src/it/java/io/sphere/sdk/types/commands/TypeCreateCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package io.sphere.sdk.types.commands;
+
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.models.TextInputHint;
+import io.sphere.sdk.test.IntegrationTest;
+import io.sphere.sdk.types.FieldDefinition;
+//important
```
Comment (lauraluiz) [inline sphere-models/src/it/java/io/sphere/sdk/types/commands/TypeCreateCommandTest.java]: Just curious, why?

File: sphere-models/src/main/java/io/sphere/sdk/carts/Cart.java [hunk 3]
```diff
@@ -79,4 +93,8 @@ public String toString() {
 
     @Override
     List<DiscountCodeInfo> getDiscountCodes();
+
+    @Nullable
+    @Override
+    CustomFields getCustom();
```
Comment (lauraluiz) [inline sphere-models/src/main/java/io/sphere/sdk/carts/Cart.java]: I know this name is coming from the platform, but do you know why is it not `CustomFields` or similar?

File: sphere-models/src/it/java/io/sphere/sdk/types/commands/TypeCreateCommandTest.java (hunk 2 above)
Comment (schleichardt) [inline sphere-models/src/it/java/io/sphere/sdk/types/commands/TypeCreateCommandTest.java]: If you use `import io.sphere.sdk.types.*` then you don't get conflicts with the product attributes. We have there classes with the same name in different packages and the IDE suggests the wrong one.

File: sphere-models/src/main/java/io/sphere/sdk/carts/Cart.java (hunk 3 above)
Comment (schleichardt) [inline sphere-models/src/main/java/io/sphere/sdk/carts/Cart.java]: No, you need to ask @yanns .

Comment (yanns) [inline sphere-models/src/main/java/io/sphere/sdk/carts/Cart.java]: I know it is kinda weird in the code, but the goal is to have a good json and predicate support:
- json:
```
{
  [...]
  "custom": {
    "fieldA": "valueA"
  }
}
```

and the predicate `myobject.custom.fieldA = "valueA" `

File: sphere-models/src/main/java/io/sphere/sdk/types/Custom.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package io.sphere.sdk.types;
+
+public interface Custom {
```
Comment (lauraluiz) [inline sphere-models/src/main/java/io/sphere/sdk/types/Custom.java]: Still not convinced about the naming. Wouldn't it be better `Customizable` or `CustomizableFields`?

Comment (yanns) [inline sphere-models/src/main/java/io/sphere/sdk/types/Custom.java]: sorry I made a mistake in my first comment.
The json is actually:
```
{
  [...],
  "custom": {
    "type": { ... },
    "fields": {
      "fieldA": "valueA"
    }
  }
}
```

Replacing `custom` with `customizable` or `customizableFields` in the JSON is not very pretty.
We designed the JSON first. In that case, the name does not apply very well to OO-languages. But it is the best solution we found.
If you want to, you can change the class name. Just be sure the serialize/deserialize it to `custom`.

**Comment by schleichardt** [comment]
I missed to test the sets.

**Pull request merged**